bin/webpack-dev-server.js

```diff
@@ -1,126 +1,151 @@
+    cli.preprocess();
+  }
+  const path = require('path');
+  const pkgPath = require.resolve(`${cli.package}/package.json`);
+  // eslint-disable-next-line import/no-dynamic-require
+  const pkg = require(pkgPath);
+  // eslint-disable-next-line import/no-dynamic-require
+  require(path.resolve(path.dirname(pkgPath), pkg.bin[cli.binName]));
+};
 
-  process.exitCode = 1;
-}
+/**
+ * @typedef {Object} CliOption
+ * @property {string} name display name
+ * @property {string} package npm package name
+ * @property {string} binName name of the executable file
+ * @property {boolean} installed currently installed?
+ * @property {string} url homepage
+ * @property {function} preprocess preprocessor
+ */
+
+/** @type {CliOption} */
+const cli = {
+  name: 'webpack-cli',
+  package: 'webpack-cli',
+  binName: 'webpack-cli',
+  installed: isInstalled('webpack-cli'),
+  url: 'https://github.com/webpack/webpack-cli',
+  preprocess() {
+    process.argv.splice(2, 0, 'serve');
+  },
+};
 
-yargs.usage(
-  `${getVersions()}\nUsage:  https://webpack.js.org/configuration/dev-server/`
-);
-
-// webpack-cli@3.3 path : 'webpack-cli/bin/config/config-yargs'
-let configYargsPath;
-try {
-  require.resolve('webpack-cli/bin/config/config-yargs');
-  configYargsPath = 'webpack-cli/bin/config/config-yargs';
-} catch (e) {
-  configYargsPath = 'webpack-cli/bin/config-yargs';
-}
-// eslint-disable-next-line import/no-extraneous-dependencies
-// eslint-disable-next-line import/no-dynamic-require
-require(configYargsPath)(yargs);
-
-// It is important that this is done after the webpack yargs config,
-// so it overrides webpack's version info.
-yargs.version(getVersions());
-yargs.options(options);
-
-const argv = yargs.argv;
-
-// webpack-cli@3.3 path : 'webpack-cli/bin/utils/convert-argv'
-let convertArgvPath;
-try {
-  require.resolve('webpack-cli/bin/utils/convert-argv');
-  convertArgvPath = 'webpack-cli/bin/utils/convert-argv';
-} catch (e) {
-  convertArgvPath = 'webpack-cli/bin/convert-argv';
-}
-// eslint-disable-next-line import/no-extraneous-dependencies
-// eslint-disable-next-line import/no-dynamic-require
-const config = require(convertArgvPath)(yargs, argv, {
-  outputFilename: '/bundle.js',
-});
+if (!cli.installed) {
+  const path = require('path');
+  const fs = require('graceful-fs');
+  const readLine = require('readline');
 
-function startDevServer(config, options) {
-  let compiler;
+  const notify = `CLI for webpack must be installed.\n  ${cli.name} (${cli.url})\n`;
 
-  const configArr = config instanceof Array ? config : [config];
-  const statsColors = getColorsOption(configArr);
+  console.error(notify);
 
-  try {
-    compiler = webpack(config);
-  } catch (err) {
-    if (err instanceof webpack.WebpackOptionsValidationError) {
-      console.error(colors.error(statsColors, err.message));
-      // eslint-disable-next-line no-process-exit
-      process.exit(1);
-    }
+  const isYarn = fs.existsSync(path.resolve(process.cwd(), 'yarn.lock'));
```

Comment:
Sorry, my mistake, let's get logic from https://github.com/webpack/webpack/pull/11945/files, we will merge it, so we will support pnpm (no big different in code) also

Reply:
https://github.com/webpack/webpack/pull/11945/files merged, we should have with the same, but with `serve` as you already done

Reply:
Done :smile: